File: api/workflow/smarttraven/chat_completion/chat_completion.py
# -*- coding: utf-8 -*-
"""
SmartTavern AI 对话补全工作流 - API 封装层

提供统一的AI对话补全能力：
- 读取对话文件获取messages
- 使用LLM配置调用AI
- 保存响应到对话文件
"""
from typing import Any, Dict
import core

from .impl import (
    chat_completion_non_streaming as _chat_completion_non_streaming,
    chat_completion_streaming as _chat_completion_streaming,
)


@core.register_api(
    path="smarttraven/chat_completion/complete",
    name="AI对话补全（非流式）",
    description="读取对话文件，调用AI生成响应，保存到对话文件。适用于一次性获取完整响应的场景",
    input_schema={
        "type": "object",
        "properties": {
            "conversation_file": {"type": "string"},
            "llm_config_file": {"type": "string"},
        },
        "required": ["conversation_file", "llm_config_file"],
        "additionalProperties": False,
    },
    output_schema={
        "type": "object",
        "properties": {
            "success": {"type": "boolean"},
            "node_id": {"type": "string"},
            "content": {"type": "string"},
            "usage": {"type": "object", "additionalProperties": True},
            "response_time": {"type": "number"},
            "model_used": {"type": "string"},
            "doc": {"type": "object", "additionalProperties": True},
            "error": {"type": "string"},
        },
        "required": ["success"],
        "additionalProperties": True,
    },
)
def complete(
    conversation_file: str,
    llm_config_file: str,
) -> Dict[str, Any]:
    return _chat_completion_non_streaming(
        conversation_file=conversation_file,
        llm_config_file=llm_config_file,
    )


@core.register_api(
    path="smarttraven/chat_completion/complete_stream",
    name="AI对话补全（流式）",
    description="读取对话文件，调用AI流式生成响应，逐块返回并最终保存到对话文件。返回SSE事件流",
    input_schema={
        "type": "object",
        "properties": {
            "conversation_file": {"type": "string"},
            "llm_config_file": {"type": "string"},
        },
        "required": ["conversation_file", "llm_config_file"],
        "additionalProperties": False,
    },
    output_schema={
        "type": "object",
        "description": "SSE流，事件类型：chunk/finish/usage/saved/error/end",
        "additionalProperties": True,
    },
)
def complete_stream(
    conversation_file: str,
    llm_config_file: str,
) -> Any:
    """
    流式补全：返回SSE
    事件格式：
    - data: {"type": "chunk", "content": "..."}
    - data: {"type": "finish", "finish_reason": "..."}
    - data: {"type": "usage", "usage": {...}}
    - data: {"type": "saved", "node_id": "...", "doc": {...}}
    - data: {"type": "error", "message": "..."}
    - data: {"type": "end"}
    """
    try:
        from fastapi.responses import StreamingResponse
    except Exception as e:
        return {
            "success": False,
            "error": f"SSE不可用（依赖fastapi未就绪）: {str(e)}"
        }
    
    import json
    
    def _sse_line(obj: Dict[str, Any]) -> str:
        return "data: " + json.dumps(obj, ensure_ascii=False) + "\n\n"
    
    def _make_sse_generator():
        try:
            for event in _chat_completion_streaming(
                conversation_file=conversation_file,
                llm_config_file=llm_config_file,
            ):
                yield _sse_line(event)
        except Exception as e:
            yield _sse_line({"type": "error", "message": str(e)})
            yield _sse_line({"type": "end"})
    
    return StreamingResponse(
        _make_sse_generator(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
            "X-Accel-Buffering": "no",
        },
    )

File: frontend_projects/SmartTavern/src/services/chatCompletion.js
// SmartTavern Frontend API Client — Chat Completion Workflow
// 调用后端 workflow.smarttraven/chat_completion 系列接口
// 默认网关：http://localhost:8050/api/workflow

const DEFAULT_BASE = 'http://localhost:8050/api/workflow';

function ensureBase() {
  const fromWindow = typeof window !== 'undefined' && window.ST_API_BASE;
  return String(fromWindow || DEFAULT_BASE).replace(/\/+$/, '');
}

async function postJSON(path, body = {}) {
  const base = ensureBase();
  const url = `${base}/${String(path).replace(/^\/+/, '')}`;
  let resp;
  try {
    resp = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body || {}),
    });
  } catch (networkError) {
    const err = new Error(`[ChatCompletion] Network error: ${networkError?.message || networkError}`);
    err.cause = networkError;
    err.url = url;
    throw err;
  }

  let data = null;
  const text = await resp.text().catch(() => '');
  try {
    data = text ? JSON.parse(text) : null;
  } catch (parseError) {
    const err = new Error(`[ChatCompletion] Invalid JSON response (${resp.status}): ${text?.slice(0, 200)}`);
    err.cause = parseError;
    err.status = resp.status;
    err.url = url;
    throw err;
  }

  if (!resp.ok) {
    const err = new Error(`[ChatCompletion] HTTP ${resp.status}: ${data && (data.message || data.error) || 'Unknown error'}`);
    err.status = resp.status;
    err.url = url;
    err.details = data;
    throw err;
  }
  return data;
}

const ChatCompletion = {
  /**
   * 非流式AI补全
   * @param {string} conversationFile - 对话文件路径
   * @param {string} llmConfigFile - LLM配置文件路径
   * @returns {Promise<Object>} { success, node_id, content, usage, response_time, model_used, doc }
   */
  async complete({ conversationFile, llmConfigFile }) {
    if (!conversationFile || !llmConfigFile) {
      throw new Error('[ChatCompletion] conversationFile and llmConfigFile are required');
    }
    
    return postJSON('smarttraven/chat_completion/complete', {
      conversation_file: conversationFile,
      llm_config_file: llmConfigFile
    });
  },

  /**
   * 流式AI补全（使用fetch + ReadableStream）
   * @param {string} conversationFile - 对话文件路径
   * @param {string} llmConfigFile - LLM配置文件路径
   * @param {Object} callbacks - 回调函数
   * @param {Function} callbacks.onChunk - 收到文本块时调用 (content: string)
   * @param {Function} callbacks.onFinish - 完成时调用 (finish_reason: string)
   * @param {Function} callbacks.onUsage - 收到用量信息时调用 (usage: object)
   * @param {Function} callbacks.onSaved - 保存成功时调用 ({ node_id, doc })
   * @param {Function} callbacks.onError - 错误时调用 (message: string)
   * @param {Function} callbacks.onEnd - 流结束时调用 ()
   * @returns {Object} { abort: Function } - 可用于取消请求
   */
  completeStream({ conversationFile, llmConfigFile, callbacks = {} }) {
    if (!conversationFile || !llmConfigFile) {
      throw new Error('[ChatCompletion] conversationFile and llmConfigFile are required');
    }

    const base = ensureBase();
    const url = `${base}/smarttraven/chat_completion/complete_stream`;

    const abortController = new AbortController();

    // 使用 fetch API 进行 POST 请求
    fetch(url, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Accept': 'text/event-stream'
      },
      body: JSON.stringify({
        conversation_file: conversationFile,
        llm_config_file: llmConfigFile
      }),
      signal: abortController.signal
    })
    .then(response => {
      if (!response.ok) {
        throw new Error(`HTTP ${response.status}`);
      }

      const reader = response.body.getReader();
      const decoder = new TextDecoder();
      let buffer = '';

      // 递归读取流
      function readChunk() {
        reader.read().then(({ done, value }) => {
          if (done) {
            return;
          }

          // 解码并添加到缓冲区
          buffer += decoder.decode(value, { stream: true });

          // 按行处理
          const lines = buffer.split('\n');
          buffer = lines.pop() || ''; // 保留最后不完整的行

          for (const line of lines) {
            if (line.startsWith('data: ')) {
              const dataStr = line.slice(6).trim();
              if (!dataStr) continue;

              try {
                const data = JSON.parse(dataStr);

                switch (data.type) {
                  case 'chunk':
                    callbacks.onChunk?.(data.content);
                    break;
                  case 'finish':
                    callbacks.onFinish?.(data.finish_reason);
                    break;
                  case 'usage':
                    callbacks.onUsage?.(data.usage);
                    break;
                  case 'saved':
                    callbacks.onSaved?.({
                      node_id: data.node_id,
                      doc: data.doc,
                      usage: data.usage
                    });
                    break;
                  case 'error':
                    callbacks.onError?.(data.message);
                    return; // 停止读取
                  case 'end':
                    callbacks.onEnd?.();
                    return; // 停止读取
                }
              } catch (err) {
                console.error('[ChatCompletion] Failed to parse SSE data:', err);
              }
            }
          }

          // 继续读取下一块
          readChunk();
        }).catch(err => {
          if (err.name !== 'AbortError') {
            console.error('[ChatCompletion] Stream read error:', err);
            callbacks.onError?.('Stream read error');
          }
        });
      }

      readChunk();
    })
    .catch(err => {
      if (err.name !== 'AbortError') {
        console.error('[ChatCompletion] Fetch error:', err);
        callbacks.onError?.(err.message || 'Request failed');
      }
    });

    return {
      abort: () => abortController.abort()
    };
  },
};

export default ChatCompletion;